### PR TITLE
cachi2 yarn: test cases

### PR DESCRIPTION
**Pull request by slimreaper35**
STONEBLD-1686

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [n/a] Code coverage from testing does not decrease and new code is covered
- [n/a] Docs updated (if applicable)
- [n/a] Docs links in the code are still valid (if docs were updated)

### CODE REVIEW

**Comment by slimreaper35** [comment]
Temporary repository for testing is [here](https://github.com/slimreaper35/yarnberry-tmp). I will clone that to cachito-testing.

**Comment by slimreaper35** [comment]
Testing repo is [here](https://github.com/cachito-testing/cachi2-yarn-berry).

File: tests/integration/test_yarn.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        ),
+    ],
+)
+def test_yarn_packages(
+    test_params: utils.TestParameters,
+    cachi2_image: utils.ContainerImage,
+    tmp_path: Path,
+    test_data_dir: Path,
+    request: pytest.FixtureRequest,
+) -> None:
```
Comment (ben-alkov) [inline tests/integration/test_yarn.py:46]: A docstring suggestion would be nice

**Comment by ben-alkov** [comment]
@slimreaper35; Is this a W.I.P.? i.e. will you be adding more test scenarios?

**Comment by slimreaper35** [comment]
Probably not, there is another [story](https://issues.redhat.com/browse/STONEBLD-1687) for that, which should cover e2e test as well.

**Comment by brunoapimentel** [comment]
https://github.com/cachito-testing/cachi2-yarn-berry/blob/offline-cache/package.json#L23

According to the design, we will ban any dependency with the `exec` protocol, so this would make any request targeting this branch fail. We will probably want to cover these "fail" scenarios, the question is if we want it in this PR.

This will likely also happen to Git dependencies, but I think we should keep them until this is settled.

File: tests/integration/test_yarn.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            utils.TestParameters(
+                repo="https://github.com/cachito-testing/cachi2-yarn-berry.git",
+                ref="a37898af6d5f2fa4335e179444fad1ca7ad74613",
+                packages=({"path": ".", "type": "yarn"},),
+                check_vendor_checksums=False,
+                expected_exit_code=0,
+                expected_output="All dependencies fetched successfully",
+            ),
+            marks=pytest.mark.xfail,  # temporary
+            id="zero-installs",
```
Comment (chmeliik) [inline tests/integration/test_yarn.py:36]: Let's follow the same `id` naming as other integration tests: `yarn_{test_case}` or `yarnberry_{test_case}`. This makes it so that you can run all the yarn tests with `tox -e integration -- -k yarn`

File: tests/integration/test_yarn.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            utils.TestParameters(
+                repo="https://github.com/cachito-testing/cachi2-yarn-berry.git",
+                ref="76131443f7472916b316954c957259ee4cee97a7",
+                packages=({"path": ".", "type": "yarn"},),
+                check_vendor_checksums=False,
+                expected_exit_code=0,
+                expected_output="All dependencies fetched successfully",
+            ),
+            marks=pytest.mark.xfail,  # temporary
+            id="offline-cache",
```
Comment (chmeliik) [inline tests/integration/test_yarn.py:24]: This isn't an [offline cache](https://v3.yarnpkg.com/features/offline-cache) test case

The offline cache is the .yarn/cache directory which contains all the dependencies. The second test case has the offline cache, this one doesn't. (The second one also has the .pnp.cjs file. Offline cache + .pnp.cjs = zero-installs.)

File: tests/integration/test_yarn.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+log = logging.getLogger(__name__)
+
+
+@pytest.mark.parametrize(
+    "test_params",
+    [
+        pytest.param(
+            utils.TestParameters(
+                repo="https://github.com/cachito-testing/cachi2-yarn-berry.git",
+                ref="76131443f7472916b316954c957259ee4cee97a7",
```
Comment (chmeliik) [inline tests/integration/test_yarn.py:17]: In the READMEs, could you explain which features the test cases cover?

Comment (chmeliik) [reply tests/integration/test_yarn.py:17]: The commit history might be helpful for that https://github.com/cachito-testing/cachi2-yarn-berry/commits/offline-cache

**Review by chmeliik — APPROVED** [review]
LGTM, just the description in https://github.com/cachito-testing/cachi2-yarn-berry/tree/no-zero-installs looks incomplete

> This integration test is to ensure that the regular workflow of Yarnberry.

**Comment by slimreaper35** [comment]
updated commit hash

**Comment by brunoapimentel** [comment]
Typo in "workflow": https://github.com/cachito-testing/cachi2-yarn-berry/blob/no-zero-installs/README.md?plain=1#L1

**Comment by chmeliik** [comment]
> "the-answer" does not specify a version in its package.json [file](https://github.com/cachito-testing/cachi2-yarn-berry/blob/zero-installs/packages/the-answer/package.json), which means that `yarn info -AR --json --cache` resolves it as `0.0.0`.
> 
> Is this an intentional test case? (I think it should be 😅)

I did the bare minimum when preparing the tests, and for yarn the that does not include setting a `version` (so yeah, I think we do want to test that, because users will likely do the bare minimum as well)

**Comment by chmeliik** [comment]
> Maybe it would be nice to have in the README of the test repo a brief note on all scenarios we're trying to cover here, just like we have for the [protocols](https://github.com/cachito-testing/cachi2-yarn-berry/blob/no-zero-installs/README.md?plain=1#L6). Some other things I noticed we also have in the test repo:
> 
>     * dev, optional and peer dependencies
> 
>     * workspaces
> 
>     * scoped dependencies
> 
>     * [pre|post]install scripts? (I see them in the no-zero-installs package.json)
> 
> 
> @chmeliik anything else worth noting?

Maybe these?

* https://github.com/cachito-testing/cachi2-yarn-berry/commit/4326fbbde1b1770e752138a9347e13fcfaabc9be
  * dependency not relevant for the current platform, we want to fetch it anyway
* https://github.com/cachito-testing/cachi2-yarn-berry/commit/4066baa6ff91ce6d4491370479d8b8c23ed7dd37
  * dependency that has to be compiled (the commit message says postinstall script, I think that's not accurate - I think yarnberry knows that compilation is needed based on the presence of [binding.gyp](https://github.com/tree-sitter/tree-sitter-json/blob/master/binding.gyp))

**Review by ejegrova — APPROVED** [review]

**Comment by lkolacek** [comment]
Everything looks great :+1:, but I have 1 (maybe stupid) question. Project requested yarn 3.6.1 and `yarn install` on node18 container changed it to 3.6.3. Is this expected behavior? I would expect corepack to use 3.6.1 version which seems to be valid. Haven't found any useful info in yarn install log about this change.

**Comment by lkolacek** [comment]
> Everything looks great 👍, but I have 1 (maybe stupid) question. Project requested yarn 3.6.1 and `yarn install` on node18 container changed it to 3.6.3. Is this expected behavior? I would expect corepack to use 3.6.1 version which seems to be valid. Haven't found any useful info in yarn install log about this change.

My bad, I updated yarn version via corepack.

**Review by lkolacek — APPROVED** [review]

**Comment by slimreaper35** [comment]
So, what is the decision :smile: on README.md in the [repo](https://github.com/cachito-testing/cachi2-yarn-berry/tree/no-zero-installs) @everyone ?

**Comment by chmeliik** [comment]
> So, what is the decision 😄 on README.md in the [repo](https://github.com/cachito-testing/cachi2-yarn-berry/tree/no-zero-installs) @everyone ?

Let's add the things we suggested (you can probably omit "workspaces", those are already mentioned in Protocols

**Comment by slimreaper35** [comment]
@everyone PTAL - [repo](https://github.com/cachito-testing/cachi2-yarn-berry/tree/no-zero-installs) 
---
(rebased and updated commit hashes)

**Review by chmeliik — APPROVED** [review]

**Review by brunoapimentel — APPROVED** [review]
